handlers/infobip/infobip.go

```diff
@@ -1,3 +1,124 @@
 package infobip
 
+import (
+	"errors"
+	"net/http"
+	"time"
+
+	"github.com/nyaruka/courier"
+	"github.com/nyaruka/courier/handlers"
+	"github.com/nyaruka/gocommon/urns"
+)
+
 /* no logs */
+
+func init() {
+	courier.RegisterHandler(NewHandler())
+}
+
+type handler struct {
+	handlers.BaseHandler
+}
+
+// NewHandler returns a new Infobip handler
+func NewHandler() courier.ChannelHandler {
+	return &handler{handlers.NewBaseHandler(courier.ChannelType("IB"), "Infobip")}
+}
+
+// Initialize is called by the engine once everything is loaded
+func (h *handler) Initialize(s courier.Server) error {
+	h.SetServer(s)
+	err := s.AddReceiveMsgRoute(h, "POST", "receive", h.ReceiveMessage)
+	return err
+}
+
+// ReceiveMessage is our HTTP handler function for incoming messages
+func (h *handler) ReceiveMessage(channel courier.Channel, w http.ResponseWriter, r *http.Request) ([]courier.ReceiveEvent, error) {
+	ie := &infobipEnvelope{}
+	err := handlers.DecodeAndValidateJSON(ie, r)
+	if err != nil {
+		return nil, courier.WriteError(w, r, err)
+	}
+
+	if ie.MessageCount == 0 {
+		return nil, courier.WriteIgnored(w, r, "ignoring request, no message")
+	}
+
+	msgs := []courier.Msg{}
+	for _, infobipMessage := range ie.Results {
+		messageID := infobipMessage.MessageID
+		text := infobipMessage.Text
+		dateString := infobipMessage.ReceivedAt
+
+		if text == "" {
+			continue
+		}
+
+		date := time.Now()
+		if dateString != "" {
+			date, err = time.Parse("2006-01-02T15:04:05.999999999-0700", dateString)
+			if err != nil {
+				return nil, courier.WriteError(w, r, err)
+			}
+		}
+
+		// create our URN
+		urn := urns.NewTelURNForCountry(infobipMessage.From, channel.Country())
+
+		// build our infobipMessage
+		msg := h.Backend().NewIncomingMsg(channel, urn, text).WithReceivedOn(date).WithExternalID(messageID)
+
+		// and write it
+		err = h.Backend().WriteMsg(msg)
+		if err != nil {
+			return nil, courier.WriteError(w, r, err)
```

Comment:
In this case, this is a courier error, not a validation error, so we should return err. (that will get logged to sentry and the caller will take care of returning a formatted JSON error)

Reply:
Okay I get that now